include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,591 @@
+        : pool_{other.pool_},
+          stream_{other.stream_},
+          buffers_{std::move(other.buffers_)} {}
+
+    inline Workspace& operator=(Workspace&& other) {
+      if (pool_) {
+        pool_->put_raw(buffers_.begin(), buffers_.end(), stream_);
+      }
+      pool_ = other.pool_;
+      stream_ = other.stream_;
+      buffers_ = std::move(other.buffers_);
+      other.pool_ = nullptr;
+      return *this;
+    }
+
+    inline ~Workspace() {
+      if (pool_) {
+        pool_->put_raw(buffers_.begin(), buffers_.end(), stream_);
+      }
+    }
+
+    template <class T>
+    constexpr void at(const size_t n, T* ptr) const {
+      *ptr = at<T>(n);
+    }
+
+    template <class T>
+    constexpr T at(const size_t n) const {
+      return reinterpret_cast<T>(buffers_.at(n));
+    }
+
+    template <class T>
+    constexpr void get(const size_t n, T* ptr) const {
+      *ptr = get<T>(n);
+    }
+
+    template <class T>
+    constexpr T get(const size_t n) const {
+      return reinterpret_cast<T>(buffers_[n]);
+    }
+
+    constexpr alloc_type* operator[](const size_t n) const {
+      return buffers_[n];
+    }
+
+   protected:
+    pool_type* pool_;
+    cudaStream_t stream_;
+    Container buffers_;
+  };
+
+  template <size_t N>
+  class StaticWorkspace final : public Workspace<std::array<alloc_type*, N>> {
+   public:
+    using base_type = Workspace<std::array<alloc_type*, N>>;
+
+    inline StaticWorkspace() : base_type() {}
+
+    inline StaticWorkspace(pool_type* pool, cudaStream_t stream)
+        : base_type(pool, stream) {
+      auto& buffers = this->buffers_;
+      pool->get_raw(buffers.begin(), buffers.end(), stream);
+    }
+
+    StaticWorkspace(const StaticWorkspace&) = delete;
+
+    StaticWorkspace& operator=(const StaticWorkspace&) = delete;
+
+    inline StaticWorkspace(StaticWorkspace&& other)
+        : base_type(std::move(other)) {}
+
+    inline StaticWorkspace& operator=(StaticWorkspace&& other) {
+      base_type::operator=(std::move(other));
+      return *this;
+    }
+  };
+
+  class DynamicWorkspace final : public Workspace<std::vector<alloc_type*>> {
+   public:
+    using base_type = Workspace<std::vector<alloc_type*>>;
+
+    inline DynamicWorkspace() : base_type() {}
+
+    inline DynamicWorkspace(pool_type* pool, size_t n, cudaStream_t stream)
+        : base_type(pool, stream) {
+      auto& buffers = this->buffers_;
+      buffers.resize(n);
+      pool->get_raw(buffers.begin(), buffers.end(), stream);
+    }
+
+    DynamicWorkspace(const DynamicWorkspace&) = delete;
+
+    DynamicWorkspace& operator=(const DynamicWorkspace&) = delete;
+
+    inline DynamicWorkspace(DynamicWorkspace&& other)
+        : base_type(std::move(other)) {}
+
+    inline DynamicWorkspace& operator=(DynamicWorkspace&& other) {
+      base_type::operator=(std::move(other));
+      return *this;
+    }
+  };
+
+  const MemoryPoolOptions options;
+
+  MemoryPool(const MemoryPoolOptions& options) : options{options} {
```

Comment:
Is there sort of a naming overlap?

Reply:
It is not in this context. Will work. But alas... you know what. My feelings are not strong enough to fight this one.

Let's do what you suggested before. Let's make it private, so we can append a `_`. Then the ambiguity is resolved, and you get what you wanted in your other comment (making it private).

Reply:
Done. Please resolve if happy.